### PR TITLE
EZP-26494: Error on indexing big content in TextBlock

### PR DESCRIPTION
**Pull request by andrerom**
> Fixes [EZP-26494](https://jira.ez.no/browse/EZP-26494)

### CODE REVIEW

**Comment by andrerom** [comment]
@pspanja still failing, tried to look in `TextBlockIntegrationTest` but didn't see anything obvious that needs adjustment for this. Is TextField maybe not supoorted by Solr/ES yet? Some missing mapping?

**Comment by pspanja** [comment]
@andrerom This might be because text field will be analyzed, and tests expect it to be indexed as a single keyword. I think this was the original problem, where the indexed keyword was too long.

**Comment by pspanja** [comment]
Thought a bit about this, we break the field limit here and I don't think it makes sense to index one big keyword for this field type. Full text indexing makes sense though.

So how about we take the same approach as with `RichText` field type, indexing shortened text for `Field` criterion/sort clause, and index full data for full text search?

**Comment by andrerom** [comment]
Make sense to me, so then we continue using StringField. And we rather:
- here we instead return first line, afaik separated by new line, like RichText returns first paragraph
- also fix this in Solr so string is trimmed if longer then X, in case first line or first paragraph is very long for instance (?)

**Comment by pspanja** [comment]
This ended up fixing a bit more things than initially expected.

The reported problem comes from the fact that `TextBlock` field type can store large amount of text. It doesn't really make sense to index the whole value as a single keyword, and that is where the original problem comes from:

```
Exception writing document id content94frefr to the index; possible analysis error.
...
Caused by..: java.lang.IllegalArgumentException: Document contains at least one immense term in field=\\"annals_number_txt_summary_value_s\\" (whose UTF8 encoding is longer than the max length 32766)
```

The approach taken for that problem is the same as with `RichText`: we index the field containing only shortened snippet of the whole value, and the whole value is indexed for the full text search.
Shortened snipped in this case means the following: 256 characters of the first non empty line.

Since we don't need to support Legacy admin UI anymore, the field type is also here changed to be searchable. To enable that with Legacy Search Engine, we now index (store) short text snippet with Legacy Stoarge Engine. Doing that also made possible to remove searchable override implementation used for testing Solr and Elasticsearch search engines.

Ready for review, ping @andrerom @alongosz

File: eZ/Publish/Core/FieldType/TextBlock/SearchField.php [hunk 1]
```diff
@@ -45,6 +45,18 @@ public function getIndexData(Field $field, FieldDefinition $fieldDefinition)
...
     /**
+     * Extracts short snippet of the given $string.
+     *
+     * @param string $string
+     *
+     * @return string
+     */
+    private function extractShortText($string)
+    {
+        return substr(strtok(trim($string), "\r\n"), 0, 255);
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/TextBlock/SearchField.php]: I would rather go for `mb_substr` just to be safe.

File: eZ/Publish/Core/FieldType/TextBlock/Type.php [hunk 2]
```diff
@@ -119,11 +119,14 @@ protected function checkValueStructure(BaseValue $value)
...
      * @param \eZ\Publish\Core\FieldType\TextBlock\Value $value
      *
-     * @return array
+     * @return string
      */
     protected function getSortInfo(BaseValue $value)
     {
-        return false;
+        return $this->transformationProcessor->transformByGroup(
+            substr(strtok(trim($value->text), "\r\n"), 0, 255),
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/TextBlock/Type.php]: Same comment as above - `mb_substr`

File: eZ/Publish/Core/FieldType/TextBlock/SearchField.php (hunk 1 above)
Comment (pspanja) [inline eZ/Publish/Core/FieldType/TextBlock/SearchField.php]: Good point, fixed in https://github.com/ezsystems/ezpublish-kernel/pull/1858/commits/1777d726e8e76f48369dd902d3d443abd9f2530e

File: eZ/Publish/Core/FieldType/TextBlock/Type.php (hunk 2 above)
Comment (pspanja) [inline eZ/Publish/Core/FieldType/TextBlock/Type.php]: Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/1858/commits/1777d726e8e76f48369dd902d3d443abd9f2530e

**Comment by andrerom** [comment]
+1

BUT: Composer 1.3.0 is in rc and should fix the COMPOSER_ROOT_VERSION issue and maybe also the sudden increase in memory usage _(but that might come form sudden increase in dependencies of our third party packages as well)_. So we can merge this, but we should do a followup when it is back to see if we can get rid of those things.

**Comment by pspanja** [comment]
OK, then I will keep that for 6.4, 6.5 and 6.6 and we will remove it later when fixed.

**Pull request merged**